src/troute-nwm/src/nwm_routing/__main__.py

```diff
@@ -714,6 +1403,322 @@ def main_v03(argv):
+                            data    = np.nan, 
+                            index   = network.reservoir_usgs_param_df.index, 
+                            columns = [t0]
+                        )
+
+                # USACE Reservoirs   
+                if 3 in network._waterbody_types_df['reservoir_type'].unique():
+                    if network.reservoir_usace_df.empty and len(network.reservoir_usace_param_df.index) > 0:
+                        network.reservoir_usace_df = pd.DataFrame(
+                            data    = np.nan, 
+                            index   = network.reservoir_usace_param_df.index, 
+                            columns = [t0]
+                        )
+
+                # update RFC lookback hours if there are RFC-type reservoirs in the simulation domain
+                if 4 in network._waterbody_types_df['reservoir_type'].unique():    
+                    waterbody_parameters = update_lookback_hours(dt, nts, waterbody_parameters)     
+
+            if showtiming:
+                forcing_end_time = time.time()
+                task_times['forcing_time'] += forcing_end_time - route_end_time
+  
+            if showtiming:
+                ic_end_time = time.time()
+                task_times['initial_condition_time'] += ic_end_time - forcing_end_time
+
+        if showtiming:
+            ic_start_time = time.time()
+        
+        # if streamflow DA is ON, then create a new lastobs dataframe
+        if data_assimilation_parameters:
+            streamflow_da = data_assimilation_parameters.get('streamflow_da',False)
+            if streamflow_da:
+                if streamflow_da.get('streamflow_nudging', False):
+                    network.lastobs_df = new_lastobs(run_results, dt * nts)
+            
+        if showtiming:
+            ic_end_time = time.time()
+            task_times['initial_condition_time'] += ic_end_time - ic_start_time
+        
+        nwm_output_generator(
+            run,
+            run_results,
+            supernetwork_parameters,
+            output_parameters,
+            parity_parameters,
+            restart_parameters,
+            parity_sets[run_set_iterator] if parity_parameters else {},
+            qts_subdivisions,
+            compute_parameters.get("return_courant", False),
+            cpu_pool,
+            network._waterbody_df,
+            network._waterbody_types_df,
+            data_assimilation_parameters,
+            network.lastobs_df,
+            network.link_gage_df,
+            network.link_lake_crosswalk,
+        )
+        
+        if showtiming:
+            output_end_time = time.time()
+            task_times['output_time'] += output_end_time - ic_end_time
+            
+    if showtiming:
+        task_times['total_time'] = time.time() - main_start_time
+
+    LOG.debug("process complete in %s seconds." % (time.time() - main_start_time))
+    
+    if showtiming:
+        print('************ TIMING SUMMARY ************')
```

Comment:
Since task_times contains all the relevant info, I would pull this logic into a function and pass in `task_times`